File: demo/services/metrics.py
"""A steady low-volume heartbeat with unicode sparklines.

Quiet enough to leave pinned in a second pane while you work on something else,
and a check that block-drawing characters measure correctly in the pane.
"""

import os
import random
import sys
import time

sys.path.insert(0, os.path.dirname(__file__))
from common import CYAN, GREEN, GREY, RED, RESET, YELLOW, install_clean_exit, log  # noqa: E402

BLOCKS = "▁▂▃▄▅▆▇█"


def spark(values):
    lo, hi = min(values), max(values)
    span = (hi - lo) or 1
    return "".join(BLOCKS[int((v - lo) / span * (len(BLOCKS) - 1))] for v in values)


def main():
    install_clean_exit()
    log("INFO", "metrics collector starting")
    log("INFO", f"{GREEN}ready{RESET} - scrape interval 2s")

    cpu, rps = [random.uniform(8, 40) for _ in range(10)], [random.uniform(20, 90) for _ in range(10)]
    while True:
        cpu = cpu[1:] + [max(1, min(99, cpu[-1] + random.gauss(0, 9)))]
        rps = rps[1:] + [max(1, rps[-1] + random.gauss(0, 14))]
        mem = random.uniform(280, 720)

        tint = RED if cpu[-1] > 80 else (YELLOW if cpu[-1] > 55 else GREEN)
        log(
            "INFO",
            f"cpu {tint}{cpu[-1]:5.1f}%{RESET} {GREY}{spark(cpu)}{RESET} "
            f"rps {CYAN}{rps[-1]:5.1f}{RESET} {GREY}{spark(rps)}{RESET} "
            f"mem {mem:3.0f}M",
        )
        if cpu[-1] > 85:
            log("WARN", f"{YELLOW}cpu saturation - scheduler queue depth rising{RESET}")
        time.sleep(2)


if __name__ == "__main__":
    main()

File: demo/services/metrics.ts
/**
 * A steady low-volume heartbeat with unicode sparklines.
 *
 * Quiet enough to leave pinned in a second pane while you work on something
 * else, and a check that block-drawing characters measure correctly in a pane.
 */

import { CYAN, GREEN, GREY, RED, RESET, YELLOW, gauss, installCleanExit, log, sleep } from "./common.ts";

const BLOCKS = "▁▂▃▄▅▆▇█";

function spark(values: readonly number[]): string {
  const lo = Math.min(...values);
  const hi = Math.max(...values);
  const span = hi - lo || 1;
  return values
    .map((v) => BLOCKS[Math.floor(((v - lo) / span) * (BLOCKS.length - 1))])
    .join("");
}

const clamp = (v: number, lo: number, hi: number): number => Math.max(lo, Math.min(hi, v));

async function main(): Promise<void> {
  installCleanExit();
  log("INFO", "metrics collector starting");
  log("INFO", `${GREEN}ready${RESET} - scrape interval 2s`);

  let cpu = Array.from({ length: 10 }, () => 8 + Math.random() * 32);
  let rps = Array.from({ length: 10 }, () => 20 + Math.random() * 70);

  for (;;) {
    cpu = [...cpu.slice(1), clamp(cpu[cpu.length - 1]! + gauss(0, 9), 1, 99)];
    rps = [...rps.slice(1), Math.max(1, rps[rps.length - 1]! + gauss(0, 14))];
    const mem = 280 + Math.random() * 440;
    const now = cpu[cpu.length - 1]!;

    const tint = now > 80 ? RED : now > 55 ? YELLOW : GREEN;
    log(
      "INFO",
      `cpu ${tint}${now.toFixed(1).padStart(5)}%${RESET} ${GREY}${spark(cpu)}${RESET} ` +
        `rps ${CYAN}${rps[rps.length - 1]!.toFixed(1).padStart(5)}${RESET} ${GREY}${spark(rps)}${RESET} ` +
        `mem ${String(Math.round(mem)).padStart(3)}M`,
    );
    if (now > 85) log("WARN", `${YELLOW}cpu saturation - scheduler queue depth rising${RESET}`);
    await sleep(2);
  }
}

await main();
